hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/cache/CacheManager.java

```diff
@@ -85,141 +85,150 @@ public void init() {
                 publishCache.clear();
             }
             msgLog.infoInitPublishCache();
-            requestCacheUpdate();
 
-            definitions.registerListener(e -> {
-                requestCacheUpdate();
+            initialCacheUpdate();
+
+            definitions.registerListener(events -> {
+                events.stream().forEach(e -> {
+                    String tenantId = e.getTargetTenantId();
+                    String triggerId = e.getTargetId();
+                    TriggerKey triggerKey = new TriggerKey(tenantId, triggerId);
+                    Set<String> oldDataIds = dataIdsCache.get(triggerKey);
+                    removePublishCache(tenantId, triggerId, oldDataIds);
+                    if (e.getType().equals(TRIGGER_CONDITION_CHANGE)) {
+                        Set<String> newDataIds = e.getDataIds();
+                        dataIdsCache.put(triggerKey, newDataIds);
+                        addPublishCache(tenantId, triggerId, newDataIds);
+                    }
+                });
             }, TRIGGER_CONDITION_CHANGE, TRIGGER_REMOVE);
 
+
         } else {
             msgLog.warnDisabledPublishCache();
         }
     }
 
-    // Just run updateCache one time if multiple requests come in while an update is already in progress...
-    private void requestCacheUpdate() {
-        log.debug("Cache update requested");
-        if (updateRequested) {
-            log.debug("Cache update, redundant request ignored.");
-            return;
+    private void removePublishCache(String tenantId, String triggerId, Set<String> dataIds) {
+        if (!isEmpty(dataIds)) {
+            dataIds.stream().forEach(dataId -> {
+                CacheKey cacheKey = new CacheKey(tenantId, dataId);
+                Set<String> triggers = publishCache.get(cacheKey);
+                if (!isEmpty(triggers)) {
+                    triggers.remove(triggerId);
+                    if (triggers.isEmpty()) {
+                        publishCache.remove(cacheKey);
+                    } else {
+                        publishCache.put(cacheKey, triggers);
```

Comment:
Yes.
It is key of the design.
It shows that dataId1 can be referenced on triggerId1 and triggerId2.
When dataId1 references to an empty set it means that the dataId is not longer active so it should be removed.
Also if we delete the reference from triggerId2, the triggersIds should be updated to count down the references of the dataId.
(If I don't have introduce any issue in the logic).